File: python/evaluators/llm.py
from evaluators.utils import _run_evaluator, _arun_evaluator
from evaluators.types import (
    EvaluatorResult,
    SimpleEvaluator,
    SimpleAsyncEvaluator,
    RunnableLike,
    ModelClient,
    ChatCompletionMessage,
    FewShotExample,
)

from langchain.chat_models import init_chat_model
from langsmith import traceable

from typing import (
    Callable,
    Optional,
    Union,
    TYPE_CHECKING,
)

import json

if TYPE_CHECKING:
    from langchain_core.language_models.chat_models import BaseChatModel


def _append_few_shot_examples(
    *,
    messages: list[ChatCompletionMessage],
    few_shot_examples: list[FewShotExample],
) -> list[ChatCompletionMessage]:
    # Find the last user message to append examples to
    last_user_message_idx = None
    for i, msg in enumerate(messages[::-1]):
        if msg.get("role") == "user":
            last_user_message_idx = len(messages) - 1 - i
            break

    if last_user_message_idx is None:
        raise ValueError(
            "Appending few-shot examples requires a user message in the provided prompt"
        )

    messages[last_user_message_idx]["content"] += "\n\n" + "\n".join(
        [
            f"<example>\n<input>{example['inputs']}</input>\n<output>{example['outputs']}</output>"
            + (
                f"\n<reasoning>{example['reasoning']}</reasoning>"
                if "reasoning" in example
                else ""
            )
            + (f"\n<score>{example['score']}</score>" if "score" in example else "")
            + "\n</example>"
            for example in few_shot_examples
        ]
    )
    return messages


def _construct_output_schema(
    *,
    schema: Optional[dict] = None,
    continuous: bool = False,
    choices: Optional[list[float]] = None,
    use_reasoning: bool = True,
) -> tuple[dict, str]:
    json_schema = (
        schema
        if schema is not None
        else {
            "type": "object",
            "additionalProperties": False,
            "strict": True,
        }
    )
    # Set the description for the score schema
    if choices:
        description = "A number that represents the degree to which the criteria in the prompt are met."
        score_schema = {
            "type": "number",
            "description": description,
            "enum": choices,
        }
    elif continuous:
        description = "A number that represents the degree to which the criteria in the prompt are met, from 0.0 to 1.0. 1.0 means the criteria are met perfectly. 0.0 means none of the criteria are met."
        score_schema = {
            "type": "number",
            "description": description,
        }
    else:
        description = "A score that is true if criteria in the prompt are met, and false otherwise."
        score_schema = {
            "type": "boolean",
            "description": description,
        }

    # Add reasoning if passed
    if schema is None:
        if use_reasoning:
            json_schema["properties"] = {
                "reasoning": {
                    "type": "string",
                    "description": "A human-readable explanation of the score. You MUST end the reasoning with a sentence that says: Thus, the score should be: SCORE_YOU_ASSIGN.",
                },
                "score": score_schema,
            }
            json_schema["required"] = ["reasoning", "score"]
        else:
            json_schema["properties"] = {
                "score": score_schema,
            }
            json_schema["required"] = ["score"]

    return (json_schema, description)


def _create_llm_as_judge_scorer(
    *,
    prompt: str | RunnableLike | Callable[..., list[ChatCompletionMessage]],
    system: Optional[str] = None,
    schema: Optional[dict] = None,
    judge: Optional[
        Union[
            ModelClient,
            BaseChatModel,
        ]
    ] = None,
    model: str = "openai:o3-mini",
    continuous: bool = False,
    choices: Optional[list[float]] = None,
    use_reasoning: bool = True,
    few_shot_examples: Optional[list[FewShotExample]] = None,
) -> SimpleEvaluator:
    def get_score(
        *,
        inputs: Optional[Union[str, dict]] = None,
        outputs: Union[str, dict],
        reference_outputs: Optional[Union[str, dict]] = None,
        **kwargs,
    ):
        if system is not None and not isinstance(prompt, str):
            raise ValueError(
                "`system` is only supported when `prompt` is a string template"
            )

        if isinstance(outputs, dict):
            outputs = json.dumps(outputs)
        if isinstance(reference_outputs, dict):
            reference_outputs = json.dumps(reference_outputs)
        if isinstance(inputs, dict):
            inputs = json.dumps(inputs)

        if isinstance(prompt, RunnableLike):
            formatted_prompt = prompt.invoke(
                inputs=inputs,
                outputs=outputs,
                reference_outputs=reference_outputs,
                **kwargs,
            )
            messages = formatted_prompt.messages
        elif isinstance(prompt, str):
            formatted_prompt = prompt.format(
                inputs=inputs,
                outputs=outputs,
                reference_outputs=reference_outputs,
                **kwargs,
            )
            messages = [
                {"role": "user", "content": formatted_prompt},
            ]
        else:
            messages = prompt(
                inputs=inputs,
                outputs=outputs,
                reference_outputs=reference_outputs,
                **kwargs,
            )

        if system is not None:
            messages = [
                {"role": "system", "content": system},
                *messages,
            ]

        # Add few shot examples to the prompt
        if few_shot_examples:
            messages = _append_few_shot_examples(
                messages=messages,
                few_shot_examples=few_shot_examples,
            )

        (json_schema, description) = _construct_output_schema(
            schema=schema,
            continuous=continuous,
            choices=choices,
            use_reasoning=use_reasoning,
        )

        nonlocal judge

        if judge is None:
            judge = init_chat_model(model=model)

        if isinstance(judge, BaseChatModel):
            response = judge.with_structured_output(
                {
                    "title": "score",
                    "description": description,
                    **json_schema,
                }
            ).invoke(messages)
            if schema is None:
                if use_reasoning:
                    return (response["score"], response["reasoning"])  # type: ignore
                return response["score"]  # type: ignore
            else:
                return response
        elif isinstance(judge, ModelClient):
            if model is None:
                raise ValueError("`model` is required for non-LangChain clients")
            params = {
                "messages": messages,
                "model": model,
                "response_format": {
                    "type": "json_schema",
                    "json_schema": {
                        "name": "score",
                        "strict": True,
                        "schema": json_schema,
                    },
                },
            }

            @traceable(
                run_type="llm",
                metadata={
                    "ls_provider": "openai",
                    "ls_model_name": model,
                    "ls_model_type": "chat",
                },
            )
            def invoke_llm(**params):
                return judge.chat.completions.create(**params)

            response = invoke_llm(**params)
            parsed = json.loads(response.choices[0].message.content)  # type: ignore
            if schema is None:
                if use_reasoning:
                    return (parsed["score"], parsed["reasoning"])  # type: ignore
                return parsed["score"]  # type: ignore
            else:
                return parsed
        else:
            raise ValueError("`judge` must be a ModelClient or BaseChatModel")

    return get_score


def _create_async_llm_as_judge_scorer(
    *,
    prompt: str | RunnableLike | Callable[..., list[ChatCompletionMessage]],
    system: Optional[str] = None,
    schema: Optional[dict] = None,
    judge: Optional[
        Union[
            ModelClient,
            BaseChatModel,
        ]
    ] = None,
    model: str = "openai:o3-mini",
    continuous: bool = False,
    choices: Optional[list[float]] = None,
    use_reasoning: bool = True,
    few_shot_examples: Optional[list[FewShotExample]] = None,
) -> SimpleAsyncEvaluator:
    async def aget_score(
        *,
        inputs: Optional[Union[str, dict]] = None,
        outputs: Union[str, dict],
        reference_outputs: Optional[Union[str, dict]] = None,
        **kwargs,
    ):
        if system is not None and not isinstance(prompt, str):
            raise ValueError(
                "`system` is only supported when `prompt` is a string template"
            )

        if isinstance(outputs, dict):
            outputs = json.dumps(outputs)
        if isinstance(reference_outputs, dict):
            reference_outputs = json.dumps(reference_outputs)
        if isinstance(inputs, dict):
            inputs = json.dumps(inputs)

        if isinstance(prompt, RunnableLike):
            formatted_prompt = await prompt.ainvoke(
                inputs=inputs,
                outputs=outputs,
                reference_outputs=reference_outputs,
                **kwargs,
            )
            messages = formatted_prompt.messages
        elif isinstance(prompt, str):
            formatted_prompt = prompt.format(
                inputs=inputs,
                outputs=outputs,
                reference_outputs=reference_outputs,
                **kwargs,
            )
            messages = [
                {"role": "user", "content": formatted_prompt},
            ]
        else:
            messages = prompt(
                inputs=inputs,
                outputs=outputs,
                reference_outputs=reference_outputs,
                **kwargs,
            )

        if system is not None:
            messages = [
                {"role": "system", "content": system},
                *messages,
            ]

        # Add few shot examples to the prompt
        if few_shot_examples:
            messages = _append_few_shot_examples(
                messages=messages,
                few_shot_examples=few_shot_examples,
            )

        (json_schema, description) = _construct_output_schema(
            schema=schema,
            continuous=continuous,
            choices=choices,
            use_reasoning=use_reasoning,
        )

        nonlocal judge

        if judge is None:
            judge = init_chat_model(model=model)

        if isinstance(judge, BaseChatModel):
            response = await judge.with_structured_output(
                {
                    "title": "score",
                    "description": description,
                    **json_schema,
                }
            ).ainvoke(messages)
            if schema is None:
                if use_reasoning:
                    return (response["score"], response["reasoning"])  # type: ignore
                return response["score"]  # type: ignore
            else:
                return response
        elif isinstance(judge, ModelClient):
            if model is None:
                raise ValueError("`model` is required for non-LangChain clients")
            params = {
                "messages": messages,
                "model": model,
                "response_format": {
                    "type": "json_schema",
                    "json_schema": {
                        "name": "score",
                        "strict": True,
                        "schema": json_schema,
                    },
                },
            }

            @traceable(
                run_type="llm",
                metadata={
                    "ls_provider": "openai",
                    "ls_model_name": model,
                    "ls_model_type": "chat",
                },
            )
            async def ainvoke_llm(**params):
                return await judge.chat.completions.create(**params)

            response = await ainvoke_llm(**params)
            parsed = json.loads(response.choices[0].message.content)  # type: ignore
            if schema is None:
                if use_reasoning:
                    return (parsed["score"], parsed["reasoning"])  # type: ignore
                return parsed["score"]  # type: ignore
            else:
                return parsed
        else:
            raise ValueError("`judge` must be a ModelClient or BaseChatModel")

    return aget_score


def create_llm_as_judge(
    *,
    prompt: str | RunnableLike | Callable[..., list[ChatCompletionMessage]],
    feedback_key: str = "score",
    judge: Optional[Union[ModelClient, BaseChatModel]] = None,
    model: str = "openai:o3-mini",
    system: Optional[str] = None,
    continuous: bool = False,
    choices: Optional[list[float]] = None,
    use_reasoning: bool = True,
    few_shot_examples: Optional[list[FewShotExample]] = None,
) -> SimpleEvaluator:
    """Create an evaluator that uses an LLM to assess output quality based on specified criteria.
    Handles prompt formatting, LLM invocation, and structured output parsing.

    Args:
        prompt: The evaluation prompt, can be a string template, LangChain prompt template, or callable
            that returns a list of chat messages.
        feedback_key: Key used to store the evaluation result, defaults to "score".
        judge: The LLM used for evaluation. Can be an OpenAI client),
            or a BaseChatModel. If an OpenAI client, must specify "model" as well.
            If omitted, "model" will be used to instantiate a LangChain model instance
            by model string.
        model: Model identifier to use. Defaults to "openai:o3-mini". If "judge" is an OpenAI client,
            this argument should be a model name directly. If "judge" is omitted, must be a valid
            LangChain model identifier. See `init_chat_model` docs for more details:
            https://python.langchain.com/docs/how_to/chat_models_universal_init/.
        system: Optional system message to prepend to the prompt.
        continuous: If True, score will be a float between 0 and 1. If False, score will be boolean. Defaults to False.
        choices: Optional list of specific float values the score must be chosen from.
        use_reasoning: If True, includes explanation for the score in the output. Defaults to True.
        few_shot_examples: Optional list of example evaluations to append to the prompt.

    Returns:
        A function that takes inputs, outputs, reference_outputs, and other kwargs, formats them into
        a prompt, invokes the judge, and returns an evaluation result.

    Example:
        ```python
        from openevals.evaluators.llm import create_llm_as_judge

        evaluator = create_llm_as_judge(
            prompt="Rate the quality of this response from 0 to 1: {outputs}",
            continuous=True
        )
        result = evaluator(
            inputs={"question": "What color is the sky?"},
            outputs={"response": "Blue"},
        )
        ```
    """
    scorer = _create_llm_as_judge_scorer(
        prompt=prompt,
        judge=judge,
        system=system,
        model=model,
        continuous=continuous,
        choices=choices,
        use_reasoning=use_reasoning,
        few_shot_examples=few_shot_examples,
    )

    def _wrapped_evaluator(
        *,
        inputs: dict,
        outputs: dict,
        reference_outputs: Optional[dict] = None,
        **kwargs,
    ) -> EvaluatorResult:
        run_name = (
            "llm_as_judge"
            if feedback_key == "score"
            else f"llm_as_{feedback_key}_judge"
        )
        res = _run_evaluator(
            run_name=run_name,
            scorer=scorer,
            feedback_key=feedback_key,
            inputs=inputs,
            outputs=outputs,
            reference_outputs=reference_outputs,
            **kwargs,
        )
        if isinstance(res, list):
            return res[0]
        return res  # type: ignore

    return _wrapped_evaluator  # type: ignore


def create_async_llm_as_judge(
    *,
    prompt: str | RunnableLike | Callable[..., list[ChatCompletionMessage]],
    feedback_key: str = "score",
    judge: Optional[Union[ModelClient, BaseChatModel]] = None,
    model: str = "openai:o3-mini",
    system: Optional[str] = None,
    continuous: bool = False,
    choices: Optional[list[float]] = None,
    use_reasoning: bool = True,
    few_shot_examples: Optional[list[FewShotExample]] = None,
) -> SimpleAsyncEvaluator:
    """Create an evaluator that uses an LLM to assess output quality based on specified criteria.
    Handles prompt formatting, LLM invocation, and structured output parsing.

    Args:
        prompt: The evaluation prompt, can be a string template, LangChain prompt template, or callable
            that returns a list of chat messages.
        feedback_key: Key used to store the evaluation result, defaults to "score".
        judge: The LLM used for evaluation. Can be an OpenAI client),
            or a BaseChatModel. If an OpenAI client, must specify "model" as well.
            If omitted, "model" will be used to instantiate a LangChain model instance
            by model string.
        model: Model identifier to use. Defaults to "openai:o3-mini". If "judge" is an OpenAI client,
            this argument should be a model name directly. If "judge" is omitted, must be a valid
            LangChain model identifier. See `init_chat_model` docs for more details:
            https://python.langchain.com/docs/how_to/chat_models_universal_init/.
        system: Optional system message to prepend to the prompt.
        continuous: If True, score will be a float between 0 and 1. If False, score will be boolean. Defaults to False.
        choices: Optional list of specific float values the score must be chosen from.
        use_reasoning: If True, includes explanation for the score in the output. Defaults to True.
        few_shot_examples: Optional list of example evaluations to append to the prompt.

    Returns:
        A function that takes inputs, outputs, reference_outputs, and other kwargs, formats them into
        a prompt, invokes the judge, and returns an evaluation result.

    Example:
        ```python
        from openevals.evaluators.llm import create_async_llm_as_judge

        evaluator = create_async_llm_as_judge(
            prompt="Rate the quality of this response from 0 to 1: {outputs}",
            continuous=True
        )
        result = await evaluator(
            inputs={"question": "What color is the sky?"},
            outputs={"response": "Blue"},
        )
        ```
    """
    scorer = _create_async_llm_as_judge_scorer(
        prompt=prompt,
        judge=judge,
        system=system,
        model=model,
        continuous=continuous,
        choices=choices,
        use_reasoning=use_reasoning,
        few_shot_examples=few_shot_examples,
    )

    async def _wrapped_evaluator(
        *,
        inputs: dict,
        outputs: dict,
        reference_outputs: Optional[dict] = None,
        **kwargs,
    ) -> EvaluatorResult:
        run_name = (
            "llm_as_judge"
            if feedback_key == "score"
            else f"llm_as_{feedback_key}_judge"
        )
        res = await _arun_evaluator(
            run_name=run_name,
            scorer=scorer,
            feedback_key=feedback_key,
            inputs=inputs,
            outputs=outputs,
            reference_outputs=reference_outputs,
            **kwargs,
        )
        if isinstance(res, list):
            return res[0]
        return res

    return _wrapped_evaluator  # type: ignore

File: js/src/evaluators/llm.ts
import { RunnableInterface, Runnable } from "@langchain/core/runnables";
import { BaseChatModel } from "@langchain/core/language_models/chat_models";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { BaseMessage } from "@langchain/core/messages";
import { initChatModel } from "langchain/chat_models/universal";
import { traceable } from "langsmith/traceable";

import { _runEvaluator, _normalizeToOpenAIMessagesList } from "./utils.js";
import {
  ChatCompletionMessage,
  FewShotExample,
  ModelClient,
  SingleResultScorerReturnType,
} from "./types.js";

function _isRunnableInterface(prompt: unknown): prompt is RunnableInterface {
  return Runnable.isRunnable(prompt);
}

function _isBaseChatModel(x: unknown): x is BaseChatModel {
  const model = x as BaseChatModel;
  return (
    x != null &&
    typeof x === "object" &&
    typeof model._modelType === "function" &&
    model._modelType() === "base_chat_model"
  );
}

function appendFewShotExamples({
  messages,
  fewShotExamples,
}: {
  messages: ChatCompletionMessage[];
  fewShotExamples: FewShotExample[];
}): ChatCompletionMessage[] {
  // Find the last user message to append examples to
  const lastUserMessageIdx = messages
    .slice()
    .reverse()
    .findIndex((msg) => msg.role === "user");

  if (lastUserMessageIdx === -1) {
    throw new Error(
      "Appending few-shot examples requires a user message in the provided prompt"
    );
  }

  const actualIdx = messages.length - 1 - lastUserMessageIdx;

  // eslint-disable-next-line no-param-reassign
  messages[actualIdx].content +=
    "\n\n" +
    fewShotExamples
      .map((example) => {
        let exampleStr = `<example>\n<input>${JSON.stringify(
          example.inputs
        )}</input>\n<output>${JSON.stringify(example.outputs)}</output>`;

        if (example.reasoning) {
          exampleStr += `\n<reasoning>${example.reasoning}</reasoning>`;
        }

        if (example.score !== undefined) {
          exampleStr += `\n<score>${example.score}</score>`;
        }

        exampleStr += "\n</example>";
        return exampleStr;
      })
      .join("\n");

  return messages;
}

function constructOutputSchema({
  schema,
  continuous,
  choices,
  useReasoning,
}: {
  schema?: Record<string, unknown>;
  continuous?: boolean;
  choices?: number[];
  useReasoning?: boolean;
}): [Record<string, unknown>, string] {
  const jsonSchema: Record<string, unknown> = schema ?? {
    type: "object",
    additionalProperties: false,
    strict: true,
  };

  let description: string;
  let scoreSchema: {
    type: string;
    description: string;
    enum?: number[];
  };

  if (choices) {
    description =
      "A number that represents the degree to which the criteria in the prompt are met.";
    scoreSchema = {
      type: "number",
      description,
      enum: choices,
    };
  } else if (continuous) {
    description =
      "A number that represents the degree to which the criteria in the prompt are met, from 0.0 to 1.0. 1.0 means the criteria are met perfectly. 0.0 means none of the criteria are met.";
    scoreSchema = {
      type: "number",
      description,
    };
  } else {
    description =
      "A score that is true if criteria in the prompt are met, and false otherwise.";
    scoreSchema = {
      type: "boolean",
      description,
    };
  }

  if (!schema) {
    if (useReasoning) {
      jsonSchema.properties = {
        reasoning: {
          type: "string",
          description:
            "A human-readable explanation of the score. You MUST end the reasoning with a sentence that says: Thus, the score should be: SCORE_YOU_ASSIGN.",
        },
        score: scoreSchema,
      };
      jsonSchema.required = ["reasoning", "score"];
    } else {
      jsonSchema.properties = {
        score: scoreSchema,
      };
      jsonSchema.required = ["score"];
    }
  }

  return [jsonSchema, description];
}

export const _createLLMAsJudgeScorer = (params: {
  prompt:
    | string
    | RunnableInterface
    | ((
        ...args: unknown[]
      ) => ChatCompletionMessage[] | Promise<ChatCompletionMessage[]>);
  system?: string;
  schema?: Record<string, unknown>;
  judge?: ModelClient | BaseChatModel;
  model?: string;
  continuous?: boolean;
  choices?: number[];
  useReasoning?: boolean;
  fewShotExamples?: FewShotExample[];
}) => {
  const {
    prompt,
    system,
    schema,
    model,
    continuous,
    choices,
    fewShotExamples,
  } = params;

  let judge = params.judge;
  const useReasoning = params.useReasoning ?? true;

  const getScore = async (params: {
    inputs: unknown;
    outputs: unknown;
    referenceOutputs?: unknown;
    [key: string]: unknown;
  }): Promise<SingleResultScorerReturnType> => {
    const { inputs, outputs, referenceOutputs, ...rest } = params;

    if (system && typeof prompt !== "string") {
      throw new Error(
        "`system` is only supported when `prompt` is a string template"
      );
    }

    let stringifiedInputs = inputs;
    let stringifiedOutputs = outputs;
    let stringifiedReferenceOutputs = referenceOutputs;
    if (inputs && typeof inputs !== "string") {
      stringifiedInputs = JSON.stringify(inputs);
    }
    if (outputs && typeof outputs !== "string") {
      stringifiedOutputs = JSON.stringify(outputs);
    }
    if (referenceOutputs && typeof referenceOutputs !== "string") {
      stringifiedReferenceOutputs = JSON.stringify(referenceOutputs);
    }
    const stringifiedRest = Object.fromEntries(
      Object.entries(rest).map(([key, value]) => [
        key,
        typeof value === "string" ? value : JSON.stringify(value),
      ])
    );

    let messages: (ChatCompletionMessage | BaseMessage)[] = [];

    if (_isRunnableInterface(prompt)) {
      const formattedPrompt = await prompt.invoke({
        inputs: stringifiedInputs,
        outputs: stringifiedOutputs,
        reference_outputs: stringifiedReferenceOutputs,
        ...stringifiedRest,
      });
      messages = formattedPrompt.messages;
    } else if (typeof prompt === "string") {
      const template = ChatPromptTemplate.fromTemplate(prompt);
      const formattedPrompt = await template.invoke({
        inputs: stringifiedInputs,
        outputs: stringifiedOutputs,
        reference_outputs: stringifiedReferenceOutputs,
        ...stringifiedRest,
      });
      messages = formattedPrompt.messages;
    } else {
      messages = await prompt({
        inputs,
        outputs,
        reference_outputs: referenceOutputs,
        ...rest,
      });
    }

    if (system) {
      messages = [{ role: "system", content: system }, ...messages];
    }

    let normalizedMessages = _normalizeToOpenAIMessagesList(messages);

    if (fewShotExamples) {
      normalizedMessages = appendFewShotExamples({
        messages: normalizedMessages,
        fewShotExamples,
      });
    }

    const [jsonSchema, description] = constructOutputSchema({
      schema,
      continuous,
      choices,
      useReasoning,
    });

    if (!judge) {
      judge = await initChatModel(model);
    }

    let response;
    if (_isBaseChatModel(judge)) {
      response = await judge
        .withStructuredOutput({
          title: "score",
          description,
          ...jsonSchema,
        })
        .invoke(normalizedMessages);
      if (schema === undefined) {
        if (useReasoning) {
          return [response.score, response.reasoning];
        }
        return response.score;
      } else {
        return response as SingleResultScorerReturnType;
      }
    } else {
      if (!model) {
        throw new Error("`model` is required for non-LangChain clients");
      }

      const params: Record<string, unknown> = {
        messages: normalizedMessages,
        model,
        response_format: {
          type: "json_schema",
          json_schema: {
            name: "score",
            strict: true,
            schema: jsonSchema,
          },
        },
      };

      const invokeLlm = traceable(
        judge.chat.completions.create.bind(judge.chat.completions),
        {
          metadata: {
            ls_provider: "openai",
            ls_model_name: model,
            ls_model_type: "chat",
          },
          run_type: "llm",
          name: "OpenAI Chat Completion",
        }
      );
      const response = await invokeLlm(params);
      const parsed = JSON.parse(response.choices[0].message.content as string);
      if (schema === undefined) {
        if (useReasoning) {
          return [parsed.score, parsed.reasoning];
        }
        return parsed.score;
      }
      return parsed;
    }
  };
  return getScore;
};

/**
 * Create an evaluator that uses an LLM to assess output quality based on specified criteria.
 *
 * @param params Configuration object with the following properties:
 * @param params.prompt The evaluation prompt - can be a string template, LangChain prompt template,
 *                     or function that returns a list of chat messages
 * @param params.feedbackKey Key used to store the evaluation result, defaults to "score"
 * @param params.judge The LLM used for evaluation. Can be an OpenAI client or a LangChain model.
 *                     If using OpenAI client, must specify "model" parameter.
 *                     If omitted, "model" will be used to instantiate a LangChain model instance.
 * @param params.model Model identifier to use. Defaults to "openai:o3-mini".
 *                     If "judge" is an OpenAI client, this should be a model name directly.
 *                     If "judge" is omitted, must be a valid LangChain model identifier.
 * @param params.system Optional system message to prepend to the prompt
 * @param params.continuous If true, score will be a float between 0 and 1.
 *                         If false, score will be boolean. Defaults to false.
 * @param params.choices Optional list of specific float values the score must be chosen from
 * @param params.useReasoning If true, includes explanation for the score in the output.
 *                           Defaults to true.
 * @param params.fewShotExamples Optional list of example evaluations to append to the prompt
 *
 * @returns A function that takes inputs, outputs, reference_outputs, and other kwargs,
 *          formats them into a prompt, invokes the judge, and returns an evaluation result
 *
 * @example
 * ```typescript
 * import { createLLMAsJudge } from "openevals";
 *
 * const evaluator = createLLMAsJudge({
 *   prompt: "Rate the quality of this response from 0 to 1: {outputs}",
 *   continuous: true,
 * });
 * const result = await evaluator({
 *   inputs: { question: "What color is the sky?" },
 *   outputs: { response: "Blue" },
 * });
 * ```
 */
export const createLLMAsJudge = ({
  prompt,
  feedbackKey = "score",
  model = "openai:o3-mini",
  system,
  judge,
  continuous = false,
  choices,
  useReasoning = true,
  fewShotExamples,
}: {
  prompt:
    | string
    | RunnableInterface
    | ((
        ...args: unknown[]
      ) => ChatCompletionMessage[] | Promise<ChatCompletionMessage[]>);
  feedbackKey?: string;
  model?: string;
  system?: string;
  judge?: ModelClient | BaseChatModel;
  continuous?: boolean;
  choices?: number[];
  useReasoning?: boolean;
  fewShotExamples?: FewShotExample[];
}) => {
  const scorer = _createLLMAsJudgeScorer({
    prompt,
    judge,
    model,
    system,
    continuous,
    choices,
    useReasoning,
    fewShotExamples,
  });

  const _wrappedEvaluator = async (inputs: {
    inputs: unknown;
    outputs: unknown;
    referenceOutputs?: unknown;
    [key: string]: unknown;
  }) => {
    const runName =
      feedbackKey !== "score" ? "llm_as_judge" : `llm_as_${feedbackKey}_judge`;
    return _runEvaluator(runName, scorer, feedbackKey, inputs);
  };
  return _wrappedEvaluator;
};
